### PR TITLE
Running resource predicate on kubelet.

### PR DESCRIPTION
**Pull request by jszczepkowski**
Related to #5207.

### CODE REVIEW

**Comment by jszczepkowski** [comment]
@bgrant0607 Can you please take a look? I was trying to add checking of resources and node selector predicates to kubelet. This PR is not finished yet, I just wanted to ask some questions before spending more time on it.

There were some issues with adding resource predicate. The predicate in scheduler has slightly different signature than in kubelet. In scheduler, it is taking scheduled pods and a candidate to schedule; while kubelet checks all the pods if they can run together. I was trying to factor out a function with kubelet-friendly interface from scheduler, and use it both in scheduler and kubelet. Is my approach correct?

I couldn't add node selector, as currently kubelet doesn't know which labels are assign to it. Should I add propagating of labels from master to kubelet?

**Comment by jszczepkowski** [comment]
The PR is ready for review: checking resources by kubelet is implemented (including a unit test).

**Comment by bgrant0607** [comment]
Yes, we should propagate node labels to Kubelet. More generally, Kubelet should be watching its Node object.

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1664,14 +1667,44 @@ func getHostPortConflicts(pods []api.Pod) []api.Pod {
...
+		if !conflictsMap[pod.UID] {
+			remainingPods = append(remainingPods, pod)
+		}
+	}
+	conflicts = kl.getPodsExceedingCapacity(remainingPods)
+	for _, pod := range conflicts {
+		kl.recorder.Eventf(&pod, "capacityExceeded", "Cannot start the pod due to exceeded capacity.")
+		kl.setPodStatusInCache(GetPodFullName(&pod), api.PodStatus{
+			Phase:   api.PodFailed,
+			Message: "Pod cannot be started due to exceeded capacity"})
```
Comment (bgrant0607) [inline pkg/kubelet/kubelet.go]: Ideally, would be useful to indicate which resource, and by how much (think "why pending").

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -1664,14 +1667,44 @@ func getHostPortConflicts(pods []api.Pod) []api.Pod {
...
+// handleNotfittingPods handles pods that do not fit on the node.
+// Currently conflicts on Port.HostPort values and exceeding node capacity are handled.
+func (kl *Kubelet) handleNotfittingPods(pods []api.Pod) {
 	conflicts := getHostPortConflicts(pods)
+	conflictsMap := map[types.UID]bool{}
 	for _, pod := range conflicts {
 		kl.recorder.Eventf(&pod, "hostPortConflict", "Cannot start the pod due to host port conflict.")
 		kl.setPodStatusInCache(GetPodFullName(&pod), api.PodStatus{
 			Phase:   api.PodFailed,
 			Message: "Pod cannot be started due to host port conflict"})
```
Comment (bgrant0607) [inline pkg/kubelet/kubelet.go]: I know this wasn't introduced by this PR, but ideally, it would be useful to indicate which port(s).

**Comment by bgrant0607** [comment]
Thanks. LGTM.

**Comment by bgrant0607** [comment]
Node selector and other improvements (more informative messages, etc.) can be handled in subsequent PRs.

**Comment by erictune** [comment]
It would be great to have a stand-alone integration test (like test/integration/client_test.go) where we
 - create a pod with resource limits.
 - wait to see the kubelet report that it is running (this will require a better fakeDocker maybe)
 - create a second pod with resource limits and force it onto the same kubelet (either with file based pod, or by setting Host explicitly, at pod creation, rather than in the scheduler)
 - wait to see that the kubelet reported failure of the pod, with reason indicating the lack of resources.

**Pull request merged**